### PR TITLE
fix: active dot appearing outside graph

### PR DESCRIPTION
**Pull request by cameronm-orion**
## Description

This fixes the issue of active dot and tooltip appearing outside the graph when the domain is set so that no dots are on the graph. I've added some checks to ensure that the domain is respected when rendering the active dot and tooltip.

## Related Issue

<https://github.com/recharts/recharts/issues/5086>

## Motivation and Context

This problem will occur mainly when implementing zoom and pan controls for graphs. In my application, I'm changing the domain as the user pans side to side on the graph, which led to the discovery of this edge case. 

## How Has This Been Tested?

I've added a unit test inside the ActiveDot test suite and a story in the LineChart story to verify the changes.

### Image of bug

image

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] I have added a storybook story or VR test, or extended an existing story or VR test to show my changes

## Summary by CodeRabbit

* **New Features**
  * Active dots and tooltips now respect axis domains and will remain inactive when data lies outside configured domains.

* **Improvements**
  * Active-point rendering now supports clipping so dots align visually with chart clip regions.

* **Documentation**
  * Added a story demonstrating behavior when data is excluded from the visible domain.

* **Tests**
  * Added tests validating tooltip/active-dot behavior and index resolution when data falls outside domains.

### CODE REVIEW

File: src/state/selectors/combiners/combineActiveTooltipIndex.ts [hunk 1]
```diff
@@ -1,10 +1,86 @@
...
+  axisDataKey: DataKey<unknown> | undefined,
+  domain: NumberDomain | CategoricalDomain | undefined,
+): boolean {
+  if (domain == null || axisDataKey == null) {
+    return true;
+  }
+
+  const value = getValueByDataKey(entry, axisDataKey);
+  if (value == null) {
+    return true;
```
Comment (PavelVanecek) [inline src/state/selectors/combiners/combineActiveTooltipIndex.ts:45]: How does this interact with `connectNulls` prop? https://recharts.github.io/en-US/examples/AreaChartConnectNulls/

File: src/state/selectors/combiners/combineActiveTooltipIndex.ts [hunk 2]
```diff
@@ -1,10 +1,86 @@
...
+  }
+
+  if (Array.isArray(domain) && domain.length !== 2) {
+    return true;
```
Comment (PavelVanecek) [inline src/state/selectors/combiners/combineActiveTooltipIndex.ts:49]: So in a categorical domain, this active dot outside of a chart is still going to be a problem? Can we have that fixed too?

File: src/state/selectors/combiners/combineActiveTooltipIndex.ts [hunk 3]
```diff
@@ -1,10 +1,86 @@
...
+  }
+
+  const domainValues = Array.isArray(domain) ? domain : [domain];
+  const numericValue = toFiniteNumber(value);
+
+  if (Array.isArray(domain) && domain.length === 2) {
```
Comment (PavelVanecek) [inline src/state/selectors/combiners/combineActiveTooltipIndex.ts]: You may want to reuse `isWellFormedNumberDomain` helper

File: src/state/selectors/combiners/combineActiveTooltipIndex.ts [hunk 4]
```diff
@@ -1,10 +1,86 @@
...
+    const upperBound = toFiniteNumber(domain[1]);
+    if (lowerBound != null && upperBound != null && numericValue != null) {
+      return isValueWithinNumberDomain(numericValue, [lowerBound, upperBound]);
+    }
+    if (lowerBound == null || upperBound == null) {
+      return true;
+    }
+  }
+
+  return domainValues.some(domainValue => {
```
Comment (PavelVanecek) [inline src/state/selectors/combiners/combineActiveTooltipIndex.ts]: When do we fall through to this branch? When can domain be not an array?

**Comment by PavelVanecek** [comment]
> ## [Codecov](https://app.codecov.io/gh/recharts/recharts/pull/6612?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=recharts) Report
> 
> ❌ Patch coverage is `75.96154%` with `25 lines` in your changes missing coverage. 

@cameronm-orion can you please add test coverage for the combiner file? It should be fairly simple to get to 100% there. The keyboard middleware would be nice to have tested too but it might be a bit challenging to set up.

File: src/state/selectors/combiners/combineActiveTooltipIndex.ts (hunk 1 above)
Comment (cameronm-orion) [reply src/state/selectors/combiners/combineActiveTooltipIndex.ts:45]: I made a small test in Storybook to see if this change caused issues with `connectNulls` and everything appears to work.

```tsx
const ConnectNullsTest = ({ args }: { args: Args }) => {
  const [domain, setDomain] = useState(args.domain);
  return (
    <>
      <ResponsiveContainer width="100%" height={200}>
        <LineChart {...args}>
          <CartesianGrid strokeDasharray="3 3" />
          <XAxis
            dataKey="name"
            allowDataOverflow={args.allowDataOverflow}
            type="number"
            domain={domain}
          />
          <YAxis />
          <Line type="monotone" dataKey="uv" stroke="#8884d8" fill="#8884d8" />
          <Tooltip />
          <RechartsHookInspector />
        </LineChart>
      </ResponsiveContainer>

      <ResponsiveContainer width="100%" height={200}>
        <LineChart {...args}>
          <CartesianGrid strokeDasharray="3 3" />
          <XAxis
            dataKey="name"
            allowDataOverflow={args.allowDataOverflow}
            type="number"
            domain={domain}
          />
          <YAxis />
          <Line
            connectNulls
            type="monotone"
            dataKey="uv"
            stroke="#8884d8"
            fill="#8884d8"
          />
          <Tooltip />
        </LineChart>
      </ResponsiveContainer>

      <div>
        <button
          type="button"
          onClick={() => setDomain((prev) => [prev[0] - 0.01, prev[1] - 0.01])}
        >
          Left
        </button>
        <button
          type="button"
          onClick={() => setDomain((prev) => [prev[0] + 0.01, prev[1] + 0.01])}
        >
          Right
        </button>
      </div>
    </>
  );
};

export const ConnectNullsNumbers = {
  render: (args: Args) => {
    return (
      <>
        <ConnectNullsTest args={args} />
      </>
    );
  },
  args: {
    width: 500,
    height: 200,
    domain: [1.01, 7.5],
    allowDataOverflow: true,
    margin: {
      top: 10,
      right: 30,
      left: 0,
      bottom: 0,
    },
    data: [
      { name: "1", uv: 300, pv: 456 },
      { name: "2", pv: 230 },
      { name: "3", pv: 345 },
      { name: "4", pv: 450 },
      { name: "5", pv: 321 },
      { name: "6", pv: 235 },
      { name: "7", pv: 267 },
      { name: "8", uv: 252, pv: -378 },
      { name: "9", uv: 79, pv: -210 },
      { name: "10", uv: 294, pv: -23 },
      { name: "12", uv: 43, pv: 45 },
      { name: "13", uv: -74, pv: 90 },
      { name: "14", uv: -71, pv: 130 },
      { name: "15", uv: -117, pv: 11 },
      { name: "16", uv: -186, pv: 107 },
      { name: "17", uv: -16, pv: 926 },
      { name: "18", uv: -125, pv: 653 },
      { name: "19", uv: 222, pv: 366 },
      { name: "20", uv: 372, pv: 486 },
      { name: "21", uv: 182, pv: 512 },
      { name: "22", uv: 164, pv: 302 },
      { name: "23", uv: 316, pv: 425 },
      { name: "24", uv: 131, pv: 467 },
      { name: "25", uv: 291, pv: -190 },
      { name: "26", uv: -47, pv: 194 },
      { name: "27", uv: -415, pv: 371 },
      { name: "28", uv: -182, pv: 376 },
      { name: "29", uv: -93, pv: 295 },
      { name: "30", uv: -99, pv: 322 },
      { name: "31", uv: -52, pv: 246 },
      { name: "32", uv: 154, pv: 33 },
      { name: "33", uv: 205, pv: 354 },
      { name: "34", uv: 70, pv: 258 },
      { name: "35", uv: -25, pv: 359 },
      { name: "36", uv: -59, pv: 192 },
      { name: "37", uv: -63, pv: 464 },
      { name: "38", uv: -91, pv: -2 },
      { name: "39", uv: -66, pv: 154 },
      { name: "40", uv: -50, pv: 186 },
    ],
  },
};
```

File: src/state/selectors/combiners/combineActiveTooltipIndex.ts (hunk 2 above)
Comment (cameronm-orion) [reply src/state/selectors/combiners/combineActiveTooltipIndex.ts:49]: I don't believe this bug is reproducible with categorical charts as the domain prop only works on a numerical axis, if my understanding is correct. And the domain prop is important for recreating this bug.

File: src/state/selectors/combiners/combineActiveTooltipIndex.ts (hunk 3 above)
Comment (cameronm-orion) [reply src/state/selectors/combiners/combineActiveTooltipIndex.ts]: Thanks, changed.

File: src/state/selectors/combiners/combineActiveTooltipIndex.ts (hunk 4 above)
Comment (cameronm-orion) [reply src/state/selectors/combiners/combineActiveTooltipIndex.ts]: Had a look through and this was unnecessary. I've removed it.

**Review by PavelVanecek — APPROVED** [review]
